### PR TITLE
Added loads of features via a new client called PubHtmlParsing client

### PR DESCRIPTION
**Pull request by ThinkDigitalSoftware**
This bypasses the API since it's missing lots of features, but it doesn't replace it, so all existing functionality is still present. I also updated some things that have been removed in the API.
# v3.0.3

* Fix TypeError in Dependencies.toJson
* Update dependencies

# v3.1.3
* Add PubHtmlParsingClient
* Deprecate values no longer in use on pub.dev

# v3.1.4
* Fix error where some FullPackages were showing null scores
* Add tests for scores

# v3.1.5
* Fix error where improperly formatted pages would throw an error.
    
# v3.2.0 - 3.3.0
* Add search for PubHtmlParsingClient
    * Add filters for flutter, web and all packages in search.
    * Add sorting for:  
          1. searchRelevance - *Packages are sorted by their updated time.*
          2. overAllScore - *Packages are sorted by the overall score.*
          3. recentlyUpdated - *Packages are sorted by their updated time.*
          4. newestPackage - *Packages are sorted by their created time.*
          5. popularity - *Packages are sorted by their popularity score.*

    * Add advanced search options available on pub.dev

# v3.3.3
 BREAKING CHANGE: Version.version has been renamed to Version.semanticVersion.
 
* Fix: bad state no element error when getting packages.
* Fix: versions page no longer showing up as part of tabs.

# v3.3.8
* Add sorting and filtering to PubHtmlParsingClient.getPageOfPackages.

# v3.4.0
*  Add homepageUrl, repositoryUrl, apiReferenceUrl, issuesUrl to FullPackage

# v3.4.1
*  Fix: packageUrl showing an extra prefix on some packages

# v3.4.2
*  Feat: Add support for Dart library entries.

### CODE REVIEW

**Comment by ThinkDigitalSoftware** [comment]
New tests need to be properly written ( I suck at that) and old tests need to be updated since the API has changed since then and it's not based on live results.

**Comment by ThinkDigitalSoftware** [comment]
Tests are failing in Travis.

**Comment by jimsimon** [comment]
Thanks for opening this.  I'm hoping to spend some time this weekend reviewing it.

**Comment by ThinkDigitalSoftware** [comment]
OK, great. I'm adding more to it 
Edit: Oh, that's right, they're being automatically added

**Comment by ThinkDigitalSoftware** [comment]
I'm glad you responded :) I wanted to contribute to the original package instead of releasing a new one.

**Comment by jimsimon** [comment]
Sorry, I have a 3 year old and a 6 month old so my time has been pretty spent for the last few months.  Things are starting to get easier, so I should have some time to review it soon.

**Comment by ThinkDigitalSoftware** [comment]
Awesome, thanks I definitely understand. Tell me what you think about the new tests. I still need to add some more, because things change quickly when parsing HTML, but the main thing is that they're testing with results from the live site, which I know is frowned upon, but I really need to know if the client is going to work with the current state of the “API”

**Comment by jimsimon** [comment]
Yeah, it's a pain to keep fake data in sync with something like this.  I
was actually considering creating another package that is under our control
and then targeting that for the tests.  I think this is a situation where
hitting live data is useful given the nature of the package under test.

**Comment by ThinkDigitalSoftware** [comment]
Well for the official pub API, it's not that much of an issue. But since my class is parsing the website, it needs to respond to multiple changes. For example, they just added Verified Publishers, which broke any of the packages that switched over to Verified Publishers. I  added better error control so that it returns nulls instead of crashing.

**Comment by ThinkDigitalSoftware** [comment]
I did some work on the PubClient class as well to update with some of the changes, but most of my work is on the `PubHtmlParsingClient`

**Comment by workerbee22** [comment]
When is this PR going to be live and documented?

**Comment by ThinkDigitalSoftware** [comment]
If you want a live one, you can check [my fork](https://github.com/ThinkDigitalSoftware/pub_client). It's more actively updated. You'll need to use the git repo as the source, since I'm trying not to release a duplicate package on pub

**Comment by jimsimon** [comment]
I'm still planning to get to this, I just have a lot on my plate at the
moment.  I'm going to try to review it tomorrow afternoon.  I have a new
computer so I need to set up a dart environment again, but that shouldn't
take too long.

**Comment by ThinkDigitalSoftware** [comment]
Ok. I'm sure I have more commits to add. This branch looks behind. My
latest update was in December

**Comment by jimsimon** [comment]
Do you mind updating this PR or opening a new one if that's easier?  My initial concern is the failing tests, but that'll likely take some work to fix and make them useful.  That work can be done later if it's too large of an effort.

**Comment by ThinkDigitalSoftware** [comment]
I believe the failing tests aren't because of me, if they're your original
tests. The api has changed since

**Comment by jimsimon** [comment]
Ah right, no worries then.  I'll fix them myself later.

**Comment by jimsimon** [comment]
Hm...odd, I just looked at your branch and it seems like this PR should be auto-updating but it's not.

**Comment by jimsimon** [comment]
Ah, nevermind I misread something.  I'll review this soon.

**Comment by ThinkDigitalSoftware** [comment]
Agreed. I need to make sure it gets in there

**Comment by ThinkDigitalSoftware** [comment]
Sounds good, thanks

File: example/main.dart [hunk 1]
```diff
@@ -1,7 +1,10 @@
 import 'package:pub_client/pub_client.dart';
 
-main() async {
+void main() async {
   var client = new PubClient();
   var package = await client.getPackage("test");
-  print(package.latest.version);
-}
\ No newline at end of file
+  var results = await client.getAllPackages();
```
Comment (jimsimon) [inline example/main.dart:6]: Might want to print out the results length or something here.

File: lib/pub_client.dart [hunk 2]
```diff
@@ -35,49 +32,58 @@ class PubClient {
...
-    var currentPage = 1;
-    var nextPageExists = true;
+    int currentPage = 1;
+    bool nextPageExists = true;
     while (nextPageExists) {
       Page page = await getPageOfPackages(currentPage);
       packages.addAll(page.packages);
       currentPage++;
-      if (page.next_url == null) {
+      print("Moving to page $currentPage");
```
Comment (jimsimon) [inline lib/pub_client.dart:45]: Probably don't want this here in production code as it's probably pretty spammy.

File: lib/pub_client.dart [hunk 3]
```diff
@@ -35,49 +32,58 @@ class PubClient {
...
-      throw new HttpException(response.statusCode, response.body);
+      throw HttpException(response.statusCode, response.body);
     }
     Page page = Page.fromJson(json.decode(response.body));
     return page;
   }
 
   Future<FullPackage> getPackage(String name) async {
     var url = "$baseApiUrl/packages/$name";
+    print(url);
```
Comment (jimsimon) [inline lib/pub_client.dart:79]: Same with this print statement

File: lib/pub_client.dart [hunk 4]
```diff
@@ -35,49 +32,58 @@ class PubClient {
...
     Response response = await client.get(url, headers: _HEADERS);
     if (response.statusCode >= 300) {
-      throw new HttpException(response.statusCode, response.body);
+      throw HttpException(response.statusCode, response.body);
     }
+    print(response.body);
```
Comment (jimsimon) [inline lib/pub_client.dart:84]: ...and this one...

File: jsons/pubspec.json [hunk 5]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (jimsimon) [inline jsons/pubspec.json:1]: Is this file and the one above still being used?  They look like test data but I don't see them referenced in the tests anywhere.

File: pubspec.yaml [hunk 6]
```diff
@@ -1,16 +1,27 @@
...
 homepage: https://github.com/jimsimon/pub_client
+
 dependencies:
-  http: ">=0.11.0 <0.13.0"
-  json_annotation: ">=1.0.0 <3.0.0"
+  http: ^0.12.0+2
+  json_annotation: ^3.0.0
+  gcloud: ^0.6.3
+
+  intl: ^0.16.0
```
Comment (jimsimon) [inline pubspec.yaml]: Are these two dependencies (gcloud and intl) being used? I don't see them referenced anywhere in the code.

**Comment by jimsimon** [comment]
Sorry it took so long to get back to this. I made a first pass at reviewing it and had a couple questions.

File: pubspec.yaml (hunk 6 above)
Comment (ThinkDigitalSoftware) [reply pubspec.yaml]: Not sure. They were there before I got there.

**Comment by gmpassos** [comment]
Time to merge this ;-P

It will be very interesting to have all of this published and a `null safety` release

**Comment by jimsimon** [comment]
I haven't done any Dart development in a long time. I mostly gave up on Dart when they stopped supporting modern web features in dart:html. I played with Flutter a bit and like it, but I don't do much native mobile development. I've moved on to Typescript for now.

That being said, I can merge this and cut a release. I just won't personally be spending much time on maintaining it. Pull requests will always be welcome though.

**Comment by jimsimon** [comment]
I ended up reverting this PR because so many tests were failing. Instead of continuing to attempt to maintain this project, it might be best if anyone interested takes a look at https://github.com/leoafarias/pub_api_client instead. I'm open to revisiting this and trying to get the tests passing if there's something missing from that package that this package supports.

**Pull request merged**